src/lender/tranche/operator/proportional.sol

```diff
@@ -0,0 +1,173 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+pragma solidity >=0.5.15 <0.6.0;
+
+import "ds-note/note.sol";
+import "tinlake-math/math.sol";
+import "tinlake-auth/auth.sol";
+
+contract TrancheLike {
+    function supply(address usr, uint currencyAmount, uint tokenAmount) public;
+    function redeem(address usr, uint currencyAmount, uint tokenAmount) public;
+    function tokenSupply() public returns (uint);
+}
+
+contract AssessorLike {
+    function calcAndUpdateTokenPrice(address tranche) public returns(uint);
+    function supplyApprove(address tranche, uint currencyAmount) public returns(bool);
+    function redeemApprove(address tranche, uint currencyAmount) public returns(bool);
+    function tokenAmountForONE() public returns(uint);
+}
+
+contract DistributorLike {
+    function balance() public;
+}
+
+contract ProportionalOperator is Math, DSNote, Auth  {
+    TrancheLike public tranche;
+    AssessorLike public assessor;
+    DistributorLike public distributor;
+
+    // lender mappings
+    // each value in a own map for gas-optimization
+    mapping (address => uint) public supplyMaximum;
+    mapping (address => uint) public currentSupplyLimit;
+
+    // expressed in totalCurrencyReturned notation
+    mapping (address => uint) public currencyRedeemed;
+
+    // expressed in totalPrincipalReturned notation
+    mapping (address => uint) public principalRedeemed;
+
+    bool public supplyAllowed  = true;
+
+    uint public totalCurrencyReturned;
+    uint public totalPrincipalReturned;
+
+    // total invested amount of currency and total supply of token
+    // contract implements a fixed supply token price of ONE
+    uint public totalTrancheVolume;
+
+    constructor(address tranche_, address assessor_, address distributor_) public {
+        wards[msg.sender] = 1;
+        tranche = TrancheLike(tranche_);
+        assessor = AssessorLike(assessor_);
+        distributor = DistributorLike(distributor_);
+    }
+
+    /// sets the dependency to another contract
+    function depend(bytes32 contractName, address addr) public auth {
+        if (contractName == "tranche") { tranche = TrancheLike(addr); }
+        else if (contractName == "assessor") { assessor = AssessorLike(addr); }
+        else if (contractName == "distributor") { distributor = DistributorLike(addr); }
+        else revert();
+    }
+
+    function file(bytes32 what, bool supplyAllowed_) public auth {
+        if(what == "supplyAllowed") {
+            supplyAllowed = supplyAllowed_;
+        }
+    }
+
+    function updateReturned(uint currencyReturned, uint principalReturned) public auth {
+        totalCurrencyReturned += currencyReturned;
```

Comment:
maybe we should overwrite it here instead of adding. In case, it is set with the wrong values we can correct it